RUMM-3403: Write migration guide for SDK v2 (#1516)
DataDog/dd-sdk-android: MIGRATION.MD

# Migration from 1.x to 2.0

The main changes introduced in SDK 2.0 compared to 1.x are:

> 1. All the products (RUM, Trace, Logs, etc.) are now extracted to the different modules. That allows you to integrate only what is needed into your application.

Reviewer: ```suggestion
1. All the products (RUM, Trace, Logs, etc.) are now extracted into different modules. That allows you to integrate only what is needed into your application.
```

> In comparison with version 1.x where all products were in a single artifact `com.datadoghq:dd-sdk-android:x.x.x`, now you need to adopt the following artifacts instead:

Reviewer 2: We might want to include OkHttp artifact too here as it's one that almost everyone will use

Author: Indeed, I completely forgot that in v1 OkHttp instrumentation doesn't live in the dedicated module.

> In comparison with the version 1.x where all the products were in the single artifact `com.datadoghq:dd-sdk-android:x.x.x`, now you need to add the following artifacts instead:

Reviewer: ```suggestion
In comparison with version 1.x where all products were in a single artifact `com.datadoghq:dd-sdk-android:x.x.x`, now you need to adopt the following artifacts instead:
```

* RUM: `com.datadoghq:dd-sdk-android-rum:x.x.x`
* Logs: `com.datadoghq:dd-sdk-android-logs:x.x.x`
* Trace: `com.datadoghq:dd-sdk-android-trace:x.x.x`

> * WebView Tracking: `com.datadoghq:dd-sdk-android-webview:x.x.x`

Reviewer 3: Shouldn't we mention the core module here?

Author: no need, all the modules re-export `core` module by declaring it as `api` dependency, so it comes automatically. This was done here https://github.com/DataDog/dd-sdk-android/pull/1479.

> **NOTE**: In case of NDK Crash Reporting and WebView Tracking usage it is needed also to add RUM and/or Logs artifacts to be able to report events to RUM and/or Logs respectively.

Reviewer: ```suggestion
**NOTE**: In case of NDK Crash Reporting and WebView Tracking usage it's also needed to add RUM and/or Logs artifacts to be able to report events to RUM and/or Logs respectively.
```

> **NOTE**: In case of NDK Crash Reporting and WebView Tracking usage it is needed also to add RUM and/or Logs artifacts to be able to report events to RUM and/or Logs respectively.

Reviewer: Aren't we missing a module for NDK crash reporting?

Author: no, dedicated module for NDK crash reporting exists in v1 as well.

Reference to the `com.datadoghq:dd-sdk-android` artifact should be removed from your Gradle buildscript, this artifact doesn't exist anymore.

> All other artifacts coordinates stay the same.

Author: It is a term used for artifacts in Java world https://maven.apache.org/pom.html#Maven_Coordinates

> 2. Support for the multiple SDK instances.

Reviewer: ```suggestion
2. Support for multiple SDK instances.
```

> Previously Datadog SDK was a singleton and only one SDK instance could exist in the application process. This was creating some difficulties for the use-cases like the usage of the SDK in 3rd party libraries.

Reviewer: ```suggestion
Previously Datadog SDK implemented a singleton and only one SDK instance could exist in the application process. This created obstacles for use-cases like the usage of the SDK by 3rd party libraries.
```

With version 2.0 we addressed this limitation:

* Now it is possible to initialize multiple instances of the SDK, associating them with a name.
* Many methods of the SDK can optionally take a SDK instance as an argument. If not provided, the call will be associated with the default (nameless) SDK instance.

Here is an example illustrating the typical usage in case of default instance vs named instance:

```kotlin
// default instance
Datadog.initialize(context, configuration, trackingConsent)
val userInfo = UserInfo(...)
Datadog.setUserInfo(userInfo)

// named instance
val namedSdkInstance = Datadog.initialize("myInstance", context, configuration, trackingConsent)
val userInfo = UserInfo(...)

> Datadog.setUserInfo(userInfo, sdkCore = namedSdkInstance)

Reviewer 3: Is it? it should be using the core registry, no?

```kotlin
Datadog.initialize("myInstance", context, configuration, trackingConsent)
val userInfo = UserInfo(...)
Datadog.setUserInfo(userInfo, sdkCore = Datadog.sdkInstance("myInstance"))
```

Author: this can be done this way as well, but on Android `Datadog.initialize` returns SDK instance, so we can use it directly.

```

> **NOTE**: SDK instance name should have the same value between application runs, because the storage path for the events is associated with it.

Reviewer: ```suggestion
**NOTE**: SDK instance name should have the same value between application runs. Storage paths for SDK events are associated with it.
```

# Migration from 1.x to 2.0

The main changes introduced in SDK 2.0 compared to 1.x are:

1. All relevant products (RUM, Trace, Logs, etc.) are now extracted into different modules. That allows you to integrate only what is needed into your application.

Whereas all products in version 1.x were contained in the single artifact `com.datadoghq:dd-sdk-android:x.x.x`, you now need to adopt the following artifacts instead:

* RUM: `com.datadoghq:dd-sdk-android-rum:x.x.x`
* Logs: `com.datadoghq:dd-sdk-android-logs:x.x.x`
* Trace: `com.datadoghq:dd-sdk-android-trace:x.x.x`
* WebView Tracking: `com.datadoghq:dd-sdk-android-webview:x.x.x`
* OkHttp instrumentation: `com.datadoghq:dd-sdk-android-okhttp:x.x.x`

**Note**: In case of NDK Crash Reporting and WebView Tracking usage it's also needed to add RUM and/or Logs artifacts to be able to report events to RUM and/or Logs respectively.

Reference to the `com.datadoghq:dd-sdk-android` artifact should be removed from your Gradle buildscript, this artifact doesn't exist anymore.

> All other artifacts Maven coordinates stay the same.

Reviewer 4: ```suggestion
All other artifacts, including Maven coordinates, stay the same.
```

Reviewer 4: I'm not sure about the intended message here, but it seems like an incomplete sentence and I guessed what you meant..

Author: I wanted to say that the Maven coordinates of all the other artifacts stay the same, so I guess the suggestion should be updated.

* `com.datadog.android.core.configuration.Credentials` class which was containing parameters mentioned above is removed.
* `logsEnabled`, `tracesEnabled`, `rumEnabled` are removed from the constructor in favour of the individual products configuration (see below).
* `crashReportsEnabled` constructor argument is removed. You can enable/disable JVM crash reporting by using `Configuration.Builder.setCrashReportsEnabled` method, by default JVM crash reporting is enabled.
* RUM, Logs, Trace products configuration methods are removed from `Configuration.Builder` in favour of the individual products configuration (see below).

`Datadog.initialize` method has `Credentials` class removed from the list of the arguments.

`com.datadog.android.plugin` package and all related classes/methods are removed.

### Logs product changes

All the classes related to the Logs product are now strictly in the `com.datadog.android.log` package.

In order to be able to use Logs product you need to import the following artifact:

```kotlin
implementation("com.datadoghq:dd-sdk-android-logs:x.x.x")
```

> Logs product can be enabled in the following way:

Reviewer 4: ```suggestion
You can enable the Logs product with the following script:
```

Reviewer 4: Not sure if you meant "script" or "command" here

Author: maybe `using the following code snippet` works better? It is not quite a script, it is a part of the application code.

```kotlin
val logsConfig = LogsConfiguration.Builder()
    ...
    .build()

Logs.enable(logsConfig)

val logger = Logger.Builder()
    ...
    .build()
```

API changes:

|`1.x`|`2.0`|
|---|---|
|`com.datadog.android.core.configuration.Configuration.Builder.setLogEventMapper`|`com.datadog.android.log.LogsConfiguration.Builder.setEventMapper`|
|`com.datadog.android.core.configuration.Configuration.Builder.useCustomLogsEndpoint`|`com.datadog.android.log.LogsConfiguration.Builder.useCustomEndpoint`|
|`com.datadog.android.log.Logger.Builder.setLoggerName`|`com.datadog.android.log.Logger.Builder.setName`|
|`com.datadog.android.log.Logger.Builder.setSampleRate`|`com.datadog.android.log.Logger.Builder.setRemoteSampleRate`|

> |`com.datadog.android.log.Logger.Builder.setDatadogLogsEnabled`|This method is removed. Use `com.datadog.android.log.Logger.Builder.setRemoteSampleRate(0f)` instead to disable sending logs to Datadog.|

Reviewer 4: ```suggestion
|`com.datadog.android.log.Logger.Builder.setDatadogLogsEnabled`|This method has been removed. Use `com.datadog.android.log.Logger.Builder.setRemoteSampleRate(0f)` instead to disable sending logs to Datadog.|
```

|`com.datadog.android.log.Logger.Builder.setServiceName`|`com.datadog.android.log.Logger.Builder.setService`|
|`com.datadog.android.log.Logger.Builder.setDatadogLogsMinPriority`|`com.datadog.android.log.Logger.Builder.setRemoteLogThreshold`|

### Trace product changes

> All the classes related to the Trace product are now strictly in the `com.datadog.android.trace` package (this means that all classes residing in `com.datadog.android.tracing` before have moved).

Reviewer 4: ```suggestion
All the classes related to the Trace product are now strictly contained in the `com.datadog.android.trace` package (this means that all classes residing in `com.datadog.android.tracing` before have moved).
```

> In order to be able to use Trace product you need to import the following artifact:

Reviewer 4: ```suggestion
To use the Trace product, import the following artifact:
```

```kotlin
implementation("com.datadoghq:dd-sdk-android-trace:x.x.x")
```

> Trace product can be enabled in the following way:

Reviewer 4: ```suggestion
Enable the Trace product with the following script:
```

```kotlin
val traceConfig = TraceConfiguration.Builder()
    ...
    .build()

Trace.enable(traceConfig)

val tracer = AndroidTracer.Builder()
    ...
    .build()

GlobalTracer.registerIfAbsent(tracer)
```

API changes:

|`1.x`|`2.0`|
|---|---|
|`com.datadog.android.core.configuration.Configuration.Builder.setSpanEventMapper`|`com.datadog.android.trace.TraceConfiguration.Builder.setEventMapper`|
|`com.datadog.android.core.configuration.Configuration.Builder.useCustomTracesEndpoint`|`com.datadog.android.trace.TraceConfiguration.Builder.useCustomEndpoint`|
|`com.datadog.android.tracing.AndroidTracer.Builder.setSamplingRate`|`com.datadog.android.trace.AndroidTracer.Builder.setSampleRate`|
|`com.datadog.android.tracing.AndroidTracer.Builder.setServiceName`|`com.datadog.android.trace.AndroidTracer.Builder.setService`|

### RUM product changes

> All the classes related to the RUM product are now strictly in the `com.datadog.android.rum` package.

Reviewer 4: ```suggestion
All classes related to the RUM product are now strictly contained in the `com.datadog.android.rum` package.
```

> In order to be able to use RUM product you need to import the following artifact:

Reviewer 4: ```suggestion
To use the RUM product, import the following artifact:
```

```kotlin
implementation("com.datadoghq:dd-sdk-android-rum:x.x.x")
```

> RUM product can be enabled in the following way:

Reviewer 4: ```suggestion
The RUM product can be enabled with the following script:
```

|`com.datadog.android.core.configuration.Configuration.Builder.setRumLongTaskEventMapper`|`com.datadog.android.rum.RumConfiguration.Builder.setLongTaskEventMapper`|
|`com.datadog.android.core.configuration.Configuration.Builder.useCustomRumEndpoint`|`com.datadog.android.rum.RumConfiguration.Builder.useCustomEndpoint`|
|`com.datadog.android.event.ViewEventMapper`|`com.datadog.android.rum.event.ViewEventMapper`|
|`com.datadog.android.core.configuration.VitalsUpdateFrequency`|`com.datadog.android.rum.configuration.VitalsUpdateFrequency`|
|`com.datadog.android.core.configuration.Configuration.Builder.trackInteractions`|`com.datadog.android.rum.RumConfiguration.Builder.trackUserInteractions`|
|`com.datadog.android.core.configuration.Configuration.Builder.disableInteractionTracking`|`com.datadog.android.rum.RumConfiguration.Builder.disableUserInteractionTracking`|
|`com.datadog.android.core.configuration.Configuration.Builder.sampleRumSessions`|`com.datadog.android.rum.RumConfiguration.Builder.setSessionSampleRate`|
|`com.datadog.android.core.configuration.Configuration.Builder.sampleTelemetry`|`com.datadog.android.rum.RumConfiguration.Builder.setTelemetrySampleRate`|

> |`com.datadog.android.rum.RumMonitor.Builder`|This class is removed. RUM monitor will be created and registered during the `Rum.enable` call.|

Reviewer 4: ```suggestion
|`com.datadog.android.rum.RumMonitor.Builder`|This class has been removed. The RUM monitor is created and registered during the `Rum.enable` call.|
```

|`com.datadog.android.rum.RumMonitor.Builder.sampleRumSessions`|`com.datadog.android.rum.RumConfiguration.Builder.setSessionSampleRate`|
|`com.datadog.android.rum.RumMonitor.Builder.setSessionListener`|`com.datadog.android.rum.RumConfiguration.Builder.setSessionListener`|
|`com.datadog.android.rum.RumMonitor.addUserAction`|`com.datadog.android.rum.RumMonitor.addAction`|
|`com.datadog.android.rum.RumMonitor.startUserAction`|`com.datadog.android.rum.RumMonitor.startAction`|
|`com.datadog.android.rum.RumMonitor.stopUserAction`|`com.datadog.android.rum.RumMonitor.stopAction`|

> |`com.datadog.android.rum.GlobalRum.registerIfAbsent`|This method is removed. RUM monitor will be created and registered during the `Rum.enable` call.|

Reviewer 4: ```suggestion
|`com.datadog.android.rum.GlobalRum.registerIfAbsent`|This method has been removed. The RUM monitor is created and registered during the `Rum.enable` call.|
```

|`com.datadog.android.rum.GlobalRum`|`com.datadog.android.rum.GlobalRumMonitor`|
|`com.datadog.android.rum.GlobalRum.addAttribute`|`com.datadog.android.rum.RumMonitor.addAttribute`|
|`com.datadog.android.rum.GlobalRum.removeAttribute`|`com.datadog.android.rum.RumMonitor.removeAttribute`|

### NDK Crash Reporting changes

The artifact name stays the same as before: `com.datadoghq:dd-sdk-android-ndk:x.x.x`

> NDK Crash Reporting can be enabled in the following way:

Reviewer 4: ```suggestion
NDK Crash Reporting can be enabled using the following setting:
```

Author: We probably need to stay consistent in the wording for the proposed application code snippets in this doc.

Reviewer 4: Updated.

```kotlin
NdkCrashReports.enable()
```

This configuration replaces the `com.datadog.android.core.configuration.Configuration.Builder.addPlugin` call used before.

**Note**: You should have RUM and/or Logs products enabled in order to receive NDK crash reports in RUM and/or Logs.

### WebView Tracking changes

The artifact name stays the same as before: `com.datadoghq:dd-sdk-android-webview:x.x.x`

> WebView Tracking can be enabled in the following way:

Reviewer 4: ```suggestion
You can enable WebView Tracking with the following setting:
```

```kotlin
WebViewTracking.enable(webView, allowedHosts)
```

**Note**: You should have RUM and/or Logs products enabled in order to receive events coming from WebView in RUM and/or Logs.

API changes:

|`1.x`|`2.0`|
|---|---|
|`com.datadog.android.webview.DatadogEventBridge`|This method became an `internal` class. Use `WebViewTracking` instead.|
|`com.datadog.android.rum.webview.RumWebChromeClient`|This class was removed. Use `WebViewTracking` instead.|
|`com.datadog.android.rum.webview.RumWebViewClient`|This class was removed. Use `WebViewTracking` instead.|

### OkHttp Tracking changes

In order to be able to use OkHttp Tracking you need to import the following artifact:

```kotlin
implementation("com.datadoghq:dd-sdk-android-okhttp:x.x.x")
```

> OkHttp instrumentation now supports the case when Datadog SDK is initialized after the OkHttp client, allowing you to create `com.datadog.android.okhttp.DatadogEventListener`, `com.datadog.android.okhttp.DatadogInterceptor`, and `com.datadog.android.okhttp.trace.TracingInterceptor` before Datadog SDK. OkHttp instrumentation will start reporting events to Datadog once Datadog SDK is initialized.

Reviewer 4: ```suggestion
OkHttp instrumentation now supports the case when Datadog SDK is initialized after the OkHttp client, allowing you to create `com.datadog.android.okhttp.DatadogEventListener`, `com.datadog.android.okhttp.DatadogInterceptor`, and `com.datadog.android.okhttp.trace.TracingInterceptor` before Datadog SDK. OkHttp instrumentation starts reporting events to Datadog once Datadog SDK is initialized.
```

> Also both `com.datadog.android.okhttp.DatadogInterceptor` and `com.datadog.android.okhttp.trace.TracingInterceptor` improve integration with remote configuration, allowing to control sampling dynamically.

Reviewer 4: ```suggestion
Also, both `com.datadog.android.okhttp.DatadogInterceptor` and `com.datadog.android.okhttp.trace.TracingInterceptor` improve the integration with remote configuration, allowing you to control sampling dynamically.
```

> In order to do that you need to provide your own implementation of the `com.datadog.android.core.sampling.Sampler` interface in the `com.datadog.android.okhttp.DatadogInterceptor`/`com.datadog.android.okhttp.trace.TracingInterceptor` constructor ; it will be queried for each request to make the sampling decision.

Reviewer 4: ```suggestion
In order to do that, you need to provide your own implementation of the `com.datadog.android.core.sampling.Sampler` interface in the `com.datadog.android.okhttp.DatadogInterceptor`/`com.datadog.android.okhttp.trace.TracingInterceptor` constructor. It is queried for each request to make the sampling decision.
```

## Using a Secondary Instance of the SDK

> Previously Datadog SDK implemented a singleton and only one SDK instance could exist in the application process. This created obstacles for use-cases like the usage of the SDK by 3rd party libraries.

Reviewer 4: ```suggestion
Previously, the Datadog SDK implemented a singleton and only one SDK instance could exist in the application process. This created obstacles for use-cases like the usage of the SDK by 3rd party libraries.
```

With version 2.0 we addressed this limitation:

> * Now it is possible to initialize multiple instances of the SDK, associating them with a name.

Reviewer 4: ```suggestion
* It is now possible to initialize multiple instances of the SDK, associating them with a name.
```

> * Many methods of the SDK can optionally take an SDK instance as an argument. If not provided, the call will be associated with the default (nameless) SDK instance.

Reviewer 4: ```suggestion
* Many methods of the SDK can optionally take an SDK instance as an argument. If not provided, the call is associated with the default (nameless) SDK instance.
```

Here is an example illustrating how to initialize a secondary core instance and enable products:

```kotlin
val namedSdkInstance = Datadog.initialize("myInstance", context, configuration, trackingConsent)
val userInfo = UserInfo(...)
Datadog.setUserInfo(userInfo, sdkCore = namedSdkInstance)

Logs.enable(logsConfig, namedSdkInstance)
val logger = Logger.Builder(namedSdkInstance)
    ...
    .build()

Trace.enable(traceConfig, namedSdkInstance)
val tracer = AndroidTracer.Builder(namedSdkInstance)
    ...
    .build()

Rum.enable(rumConfig, namedSdkInstance)
GlobalRumMonitor.get(namedSdkInstance)

NdkCrashReports.enable(namedSdkInstance)

WebViewTracking.enable(webView, allowedHosts, namedSdkInstance)
```

> **Note**: SDK instance name should have the same value between application runs. Storage paths for SDK events are associated with it.

Reviewer 4: ```suggestion
**Note**: The SDK instance name should have the same value between application runs. Storage paths for SDK events are associated with it.
```
